File: frontend/src/lib/events.ts
/**
 * Event names must match internal/app/events.go exactly.
 *
 * There are four, so duplicating them costs less than generating them. If the
 * list grows, generate it from the Go constants instead.
 */
export const EVENTS = {
  syncStarted: 'sync:started',
  syncFinished: 'sync:finished',
  syncFailed: 'sync:failed',
} as const

/**
 * Deliberately not in EVENTS: the shell subscribes to every name in there and
 * routes it to the sync reducer. This one belongs to the reading pane, which
 * subscribes to it itself.
 */
export const FIND_RESULTS_EVENT = 'find:results'

/**
 * What a find in the reading pane turned up.
 *
 * It arrives as an event rather than a return value because the counting
 * happens while the body is being served: the frame requests a document with a
 * query on it, and the number of matches falls out of building that document.
 * Asking separately would render the message twice and could disagree with
 * what is on screen.
 */
export interface FindResultsPayload {
  messageId: number
  query: string
  count: number
  /** The match the document scrolled to, after the asked-for index was wrapped. */
  current: number
}

export type ErrorClass = 'transient' | 'auth' | 'protocol' | 'permanent'

export interface SyncEventPayload {
  accountId: number
  email: string
  error?: string
  class?: ErrorClass
}

File: internal/app/events.go
package app

// Event names shared with the frontend. Keep them in step with
// frontend/src/lib/events.ts — there are four, so duplicating them is
// cheaper than generating them. If this list grows, generate it.
const (
	EventSyncStarted  = "sync:started"
	EventSyncFinished = "sync:finished"
	EventSyncFailed   = "sync:failed"
	EventFindResults  = "find:results"
)

// SyncEvent is the payload for all three sync events.
type SyncEvent struct {
	AccountID int64  `json:"accountId"`
	Email     string `json:"email"`
	// Error is empty except on sync:failed.
	Error string `json:"error,omitempty"`
	// Class is the error class name: transient, auth, protocol or permanent.
	// The UI branches on it — an auth failure needs a sign-in prompt, not the
	// spinner a transient one gets.
	Class string `json:"class,omitempty"`
}

// FindEvent reports what a find in the reading pane turned up.
//
// An event rather than a method the window calls, because the counting happens
// in the same pass that builds the document: the frame asks for a body with a
// query on it, and the number of matches falls out of serving that request. A
// separate call would mean rendering the message twice and, worse, could
// disagree with what is on screen if the two renders ever saw different bodies.
type FindEvent struct {
	MessageID int64  `json:"messageId"`
	Query     string `json:"query"`
	Count     int    `json:"count"`
	// Current is the match the document actually scrolled to, after the index
	// the window asked for was wrapped into range.
	Current int `json:"current"`
}

// Emitter delivers events to the frontend.
//
// main.go supplies a function backed by the Wails runtime; tests supply a
// recorder. This indirection is why internal/app does not import Wails at all,
// and therefore why every service method is unit-testable without a window.
type Emitter func(name string, data any)
